clang/include/clang/Sema/Sema.h

```diff
@@ -1759,6 +1759,34 @@ class Sema final {
     }
   };
 
+  /// Bitmask to contain the list of reasons a single diagnostic should be
+  /// emitted, based on its language.  This permits multiple offload systems
+  /// to coexist in the same translation unit.
+  enum class DeviceDiagnosticReason {
+    /// Diagnostic doesn't apply to anything. Included for completeness, but
+    /// should make this a no-op.
+    None = 0,
+    /// OpenMP specific diagnostic.
+    OmpDevice = 1 << 0,
+    OmpHost = 1 << 1,
+    OmpAll = OmpDevice | OmpHost,
+    /// CUDA specific diagnostics.
+    CudaDevice = 1 << 2,
+    CudaHost = 1 << 3,
+    CudaAll = CudaDevice | CudaHost,
+    /// SYCL specific diagnostic.
+    Sycl = 1 << 4,
```

Comment:
I think we do want a SYCL host vs SYCL device split here -- for example, attributes do special handling to avoid emitting diagnostics in host mode. We'd also want a `SyclAll` option. I'm fine if this gets done in a follow-up though.

Reply:
That is a good example!  I'd not thought of one at the time when writing this, but whoever comes along to implement that feature definitely can split the Sycl flag.